### DISCUSSION TITLE
mark method as throwing exception unconditionally

### DISCUSSION THREAD

**Discussion by ActivistInvestor** [dotnet/csharplang#1291] [General]
I have diagnostics methods that unconditionally throw exceptions, which I often use in place of a direct throw statement. When a method that is declared to return a non-void result contains a throw, the compiler doesn't generate an error because the method doesn't return anything.

I would like to declare or mark methods that unconditionally throw an exception, to allow them to be used in lieu of throw, and have the compiler recognize that, and not generate an error because the method does not return a result.

**Comment by ufcpp** [comment]
https://github.com/dotnet/csharplang/issues/538 ?

**Comment by aluanhaddad** [comment]
So basically a `never` type that implies that control flow cannot proceed past the point of invocation. I'm not sure if this has been proposed but if non nullable reference types are added, more and more code will begin to rely on control flow based analysis, making features like this inherently valuable.

**Comment by ActivistInvestor** [comment]
#538, I'm not sure. I'm mainly just looking for a way to avoid the need to put in an unreachable return statement in cases where the flow always leads to an exception in a called API.
